core/src/test/scala/org/apache/spark/sql/delta/DeltaSourceSuite.scala

```diff
@@ -337,6 +337,69 @@ class DeltaSourceSuite extends DeltaSourceSuiteBase with DeltaSQLCommandTest {
     }
   }
 
+  test("maxFilesPerTrigger: Trigger.AvailableNow respects read limits") {
+    withTempDir { inputDir =>
+      val deltaLog = DeltaLog.forTable(spark, inputDir)
+      (0 until 5).foreach { i =>
+        val v = Seq(i.toString).toDF
+        v.write.mode("append").format("delta").save(deltaLog.dataPath.toString)
+      }
+
+      val stream = spark.readStream
+        .format("delta")
+        .option(DeltaOptions.MAX_FILES_PER_TRIGGER_OPTION, "1")
+        .load(inputDir.getCanonicalPath)
+        .writeStream
+        .format("memory")
+        .trigger(Trigger.AvailableNow)
+        .queryName("maxFilesPerTriggerTest")
+
+      var q = stream.start()
+      try {
+        assert(q.awaitTermination(10000))
+        val progress = q.recentProgress.filter(_.numInputRows != 0)
+        assert(progress.length === 5)
+        progress.foreach { p =>
+          assert(p.numInputRows === 1)
+        }
+        checkAnswer(sql("SELECT * from maxFilesPerTriggerTest"), (0 until 5).map(_.toString).toDF)
+
+        // Restarting the stream should immediately terminate with no progress because no more data
+        q = stream.start()
+        assert(q.awaitTermination(10000))
+        assert(q.recentProgress.length === 5)
+      } finally {
+        q.stop()
+      }
+    }
+  }
+
+  test("Trigger.AvailableNow with an empty table") {
+    withTempDir { inputDir =>
+      val deltaLog = DeltaLog.forTable(spark, inputDir)
+      val df = Seq.empty[String].toDF("value")
+      df.write.mode("append").format("delta").save(deltaLog.dataPath.toString)
```

Comment:
Are you sure we are writing 0 parquet files here? It may add parquet files with 0 rows, which will not cover the case we want to test.

Reply:
We can either check the number of files or use SQL CREATE TABLE.

Reply:
In a second thought, this should be fine. No parquet files and 0 rows in parquet files have the same behavior.

Reply:
Actually the test didn't fail with an NPE (commenting out the null check) with the df.write. Changed it back to use sql just without a table name